device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/amqps/AmqpsSessionHandler.java

```diff
@@ -338,110 +338,120 @@ void openLinks()
-                        // sent once the twin links are opened.
-                        this.amqpsSessionStateCallback.onMessageAcknowledged(message, Accepted.getInstance(), this.getDeviceId());
-                        return true;
+                        // Don't send any twin messages while a twin subscription is in progress. Wait until the subscription
+                        // has been acknowledged by the service before sending it.
+                        return SendResult.SUBSCRIPTION_IN_PROGRESS;
                     }
 
-                    if (this.explicitInProgressTwinSubscriptionMessage == null)
-                    {
-                        createTwinLinks();
-                        this.explicitInProgressTwinSubscriptionMessage = transportMessage;
-                        return true;
-                    }
-                    else
-                    {
-                        log.debug("Rejecting twin subscription message because that subscription is already in progress");
-                        return false;
-                    }
-                }
-            }
-
-            for (AmqpsSenderLinkHandler senderLinkHandler : this.senderLinkHandlers)
-            {
-                if (senderLinkHandler instanceof AmqpsTelemetrySenderLinkHandler && messageType == DEVICE_TELEMETRY
-                        || senderLinkHandler instanceof AmqpsTwinSenderLinkHandler && messageType == DEVICE_TWIN
-                        || senderLinkHandler instanceof AmqpsMethodsSenderLinkHandler && messageType == DEVICE_METHODS)
-                {
-                    if (messageType == DEVICE_TWIN)
+                    for (SubscriptionType subscriptionType : this.implicitInProgressSubscriptionMessages.values())
                     {
-                        if (explicitInProgressTwinSubscriptionMessage != null)
+                        if (subscriptionType == SubscriptionType.DESIRED_PROPERTIES_SUBSCRIPTION)
                         {
                             // Don't send any twin messages while a twin subscription is in progress. Wait until the subscription
                             // has been acknowledged by the service before sending it.
-                            return false;
-                        }
-
-                        for (SubscriptionType subscriptionType : this.implicitInProgressSubscriptionMessages.values())
-                        {
-                            if (subscriptionType == SubscriptionType.DESIRED_PROPERTIES_SUBSCRIPTION)
-                            {
-                                // Don't send any twin messages while a twin subscription is in progress. Wait until the subscription
-                                // has been acknowledged by the service before sending it.
-                                return false;
-                            }
+                            return SendResult.SUBSCRIPTION_IN_PROGRESS;
```

Comment:
Knowing that there is at least one subscription of this category in progress is equally as valuable as knowing exactly how many are in progress, if that makes sense. In both cases, we need to hold off on sending twin messages until the subscriptions finish

Reply:
I see, so we're processing one message which happens to match one we already are waiting on a response for, and we know in this case to do nothing because we have a match. Yes?

Reply:
Yes